aten/src/ATen/native/cuda/Sort.cu

```diff
@@ -0,0 +1,236 @@
+  int64_t *indices_ptr;
+  if (!values.defined()) {
+    if (is_non_overlapping_and_dense) {
+      values = at::empty_strided(self.sizes(), self.strides(), self.options());
+    } else {
+      auto strides = at::infer_dense_strides(self.sizes(), self.strides());
+      values = at::empty_strided(self.sizes(), strides, self.options());
+    }
+  } else {
+    TORCH_CHECK(self_.scalar_type() == values.scalar_type(),
+      "Unexpected dtype for values, expect ", self_.scalar_type(), ", got ", values.scalar_type());
+    values.resize_as_(self);
+  }
+  if (values.strides() != self_.strides()) {
+    values_tmp = at::empty_strided(self_.sizes(), self_.strides(), self_.options());
+    values_ptr_ = values_tmp.data_ptr();
+  } else {
+    values_ptr_ = values.data_ptr();
+  }
+
+  if (!indices.defined()) {
+    if (is_non_overlapping_and_dense) {
+      indices = at::empty_strided(self.sizes(), self.strides(), self.options().dtype(kLong));
+    } else {
+      auto strides = at::infer_dense_strides(self.sizes(), self.strides());
+      indices = at::empty_strided(self.sizes(), strides, self.options().dtype(kLong));
+    }
+  } else {
+    TORCH_CHECK(kLong == indices.scalar_type(),
+      "Unexpected dtype for values, expect torch.long, got ", indices.scalar_type());
+    indices.resize_as_(self);
+  }
+  if (indices.strides() != self_.strides()) {
+    indices_tmp = at::empty_strided(self_.sizes(), self_.strides(), self_.options().dtype(kLong));
+    indices_ptr = indices_tmp.data_ptr<int64_t>();
+  } else {
+    indices_ptr = indices.data_ptr<int64_t>();
+  }
+
+  if (numel == 0) {
+    return {values, indices};
+  }
+
+  int64_t numel_or_intmax = std::min(numel, static_cast<int64_t>(std::numeric_limits<int>::max()));
+  int64_t nbatch = (numel_or_intmax / nsort) * nsort;
+  int64_t nsegments = nbatch / nsort;
+
+  auto segment_id = at::arange(nsegments, indices.options()).view({nsegments, 1}).expand({nsegments, nsort}).contiguous();
+  int64_t *segment_id_ptr = segment_id.data_ptr<int64_t>();
+  auto orig_indices = at::arange(nsort, indices.options()).repeat({nsegments});
+  int64_t *orig_indices_ptr = orig_indices.data_ptr<int64_t>();
+
+  auto tmp = at::empty({nbatch}, self_.options());
+  auto segment_id_tmp = at::empty_like(segment_id);
+  int64_t *segment_id_tmp_ptr = segment_id_tmp.data_ptr<int64_t>();
+  auto orig_indices_tmp = at::empty_like(orig_indices);
+  int64_t *orig_indices_tmp_ptr = orig_indices_tmp.data_ptr<int64_t>();
+
+  AT_DISPATCH_ALL_TYPES_AND(kHalf, self_.scalar_type(), "sort", [&]{
+    const scalar_t *self_ptr = self_.data_ptr<scalar_t>();
+    scalar_t *tmp_ptr = tmp.data_ptr<scalar_t>();
+    auto values_ptr = reinterpret_cast<scalar_t *>(values_ptr_);
+    int64_t remaining = numel;
+    while (remaining > 0) {
+      int64_t n = std::min(remaining, nbatch);
+      int64_t segment_bits = std::max<int64_t>(1L, static_cast<int64_t>(std::ceil(std::log2(n / nsort))));
+      at::cuda::cub::sort_pairs(
```

Comment:
I suspect instead of doing 2 sorts with the same key, it would be faster to do a single sort with arange as value, and then gather indices and segment ids according to permuted arange? thrust::gather hopefully should have async apis and not create device cache, if not, writing a simple gather kernel should not be awful?

Reply:
Yes, the two sort algorithm is faster. Now the new code is unconditionally faster than the old one. And this PR should be ready.

Reply:
New benchmark updated in https://github.com/pytorch/pytorch/pull/54626#issue-600072099